mage_integrations/mage_integrations/sources/paystack/streams/base.py

```diff
@@ -0,0 +1,172 @@
+import os
+import pytz
+import singer
+
+
+LOGGER = singer.get_logger()
+
+def is_selected(stream_catalog):
+    metadata = singer.metadata.to_map(stream_catalog.metadata)
+    stream_metadata = metadata.get((), {})
+
+    inclusion = stream_metadata.get('inclusion')
+
+    if stream_metadata.get('selected') is not None:
+        selected = stream_metadata.get('selected')
+    else:
+        selected = stream_metadata.get('selected-by-default')
+
+    if inclusion == 'unsupported':
+        return False
+
+    elif selected is not None:
+        return selected
+
+    return inclusion == 'automatic'
+
+
+class BasePaystackStream():
+    # GLOBAL PROPERTIES
+    TABLE = None
+    KEY_PROPERTIES = []
+    API_METHOD = 'GET'
+    REQUIRES = []
+
+    def __init__(self, config, state, catalog, client: PaystackClient):
+        self.config = config
+        self.state = state
+        self.catalog = catalog
+        self.client = client
+        self.substreams = []
+
+    def write_schema(self):
+        bookmark_properties = []
+        if REPLICATION_METHOD_INCREMENTAL == self.catalog.replication_method:
+            bookmark_properties = self.BOOKMARK_PROPERTIES
+        write_schema(
+            self.catalog.stream,
+            self.catalog.schema.to_dict(),
+            key_properties=self.catalog.key_properties or [],
+            bookmark_properties=bookmark_properties,
+            replication_method=self.catalog.replication_method,
+            unique_conflict_method=self.catalog.unique_conflict_method,
+            unique_constraints=self.catalog.unique_constraints,
+        )
+
+    def get_url(self):
+        """
+        Return the URL to hit for data from this stream.
+        """
+        raise RuntimeError("get_url not implemented!")
+
+    def get_abs_path(self, path):
+        return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)
+
+    # This overrides the transform_record method in the Fistown Analytics tap-framework package
+    def transform_record(self, record):
+        with singer.Transformer(integer_datetime_fmt="unix-seconds-integer-datetime-parsing") as tx:
+            metadata = {}
+
+            if self.catalog.metadata is not None:
+                metadata = singer.metadata.to_map(self.catalog.metadata)
+
+            return tx.transform(
+                record,
+                self.catalog.schema.to_dict(),
+                metadata
+            )
+
+    def get_stream_data(self, data):
+        return [self.transform_record(item) for item in data]
+
+    def load_data(self):
+        table = self.TABLE
+        api_method = self.API_METHOD
+        done = False
+        sync_interval_in_mins = 2
+
+        # Attempt to get the bookmark date from the state file (if one exists and is supplied).
+        LOGGER.info('Attempting to get the most recent bookmark_date for entity {}.'.format(self.ENTITY))
+        bookmark_date = get_last_record_value_for_table(self.state, table, 'bookmark_date')
```

Comment:
I wonder why we don't pass the bookmark directly to this method. Why do we fetch bookmark again?

Reply:
yeah sorry, i wanted to try to reuse the code in the chargebee sync since it's similar to paystack, so there might be some weird things like this. i'll try to polish it a bit more